### PR TITLE
Fix uninitializing fields when evaluating a cached constructor call in global initialization checker

### PR DESCRIPTION
**Pull request by EnzeXing**
This PR fixes the accessing uninitialized field errors given by the global initialization checker, where fields are uninitialized when evaluating a cached constructor call during the analysis.

### CODE REVIEW

File: tests/init-global/pos/cache-constructor.scala [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class Bar:
+    var f: Int = 0
+
+object A:
+    val b1 = new Bar()
+    val b2 = new Bar()
+    val b3 = new Bar()
+    b3.f = 1
+
```
Comment (liufengyun) [inline tests/init-global/pos/cache-constructor.scala:9]: Minor: trailing spaces in file.

File: compiler/src/dotty/tools/dotc/transform/init/Objects.scala [hunk 2]
```diff
@@ -950,7 +950,6 @@ class Objects(using Context @constructorOnly):
 
         val instance = OfClass(klass, outerWidened, ctor, args.map(_.value), envWidened)
         callConstructor(instance, ctor, args)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/init/Objects.scala:952]: I'd add a comment above `callConstructor` and refer to the test. Something like the following:

```
// The actual instance might be cached without running the constructor. 
// See tests/init-global/pos/cache-constructor.scala
```

**Review by liufengyun — APPROVED** [review]
Thank you @EnzeXing , this is a beautiful fix 🎉

**Review by olhotak — APPROVED** [review]
LGTM, pending @liufengyun 's minor suggestions.

**Pull request merged**